ambari-server/src/main/python/ambari_server/serverUtils.py

```diff
@@ -245,7 +245,7 @@ def get_ssl_context(properties, requested_protocol=None):
   :return: a permissive SSLContext or None
   """
 
-  if not is_api_ssl_enabled(properties):
+  if not is_api_ssl_enabled(properties) or not hasattr(ssl, 'SSLContext'):
```

Comment:
I have some doubts about whether this is enough to fix the problem:

1. The unit test expects the same results regardless of whether `SSLContext` is available:

https://github.com/apache/ambari/blob/11ea165b7c6d8da5462011fff10b6420b4f410c1/ambari-server/src/test/python/TestServerUtils.py#L124-L125

2. Functionality-wise: `urllib2.urlopen`'s `context` parameter was also added in Python 2.7.9, so passing `context=None` would result in `unexpected keyword argument 'context'`.

Reply:
Nice call on the TestServerUtils.py test... I missed that. :(

Reply:
Before changes:
**Python 2.7.5**
```
# python --version
Python 2.7.5

# python  -m unittest -v  TestAmbariServer.TestAmbariServer
...
----------------------------------------------------------------------
Ran 125 tests in 5.599s

OK

# python  -m unittest -v  TestServerUtils.TestServerUtils
...
----------------------------------------------------------------------
Ran 5 tests in 0.002s

OK
```

**Python 2.7.15**
```
# python --version
Python 2.7.15

# python  -m unittest -v  TestAmbariServer.TestAmbariServer
...
----------------------------------------------------------------------
Ran 125 tests in 5.544s

OK

# python  -m unittest -v  TestServerUtils.TestServerUtils
...
----------------------------------------------------------------------
Ran 5 tests in 0.001s

OK
```

**Python 2.7.6**
```
# ../Python-2.7.6/python  --version
Python 2.7.6

# ../Python-2.7.6/python -m unittest -v  TestAmbariServer.TestAmbariServer
...
======================================================================
ERROR: test_ldap_sync_ssl (TestAmbariServer.TestAmbariServer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/ambari/ambari-common/src/test/python/mock/mock.py", line 1199, in patched
    return func(*args, **keywargs)
  File "/root/ambari/ambari-server/src/test/python/TestAmbariServer.py", line 7804, in test_ldap_sync_ssl
    sync_ldap(options)
  File "/root/ambari/ambari-server/src/main/python/ambari_server/setupSecurity.py", line 405, in sync_ldap
    raise FatalException(1, err)
FatalException: "Fatal exception: Sync event creation failed. Error details: 'module' object has no attribute 'SSLContext', exit code 1"

----------------------------------------------------------------------
Ran 125 tests in 4.594s

FAILED (errors=1)

# ../Python-2.7.6/python  -m unittest -v  TestServerUtils.TestServerUtils
...
======================================================================
ERROR: test_get_ssl_context (TestServerUtils.TestServerUtils)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/ambari/ambari-common/src/test/python/mock/mock.py", line 1199, in patched
    return func(*args, **keywargs)
  File "/root/ambari/ambari-server/src/test/python/TestServerUtils.py", line 124, in test_get_ssl_context
    context = get_ssl_context(properties)
  File "/root/ambari/ambari-server/src/main/python/ambari_server/serverUtils.py", line 274, in get_ssl_context
    context = ssl.SSLContext(protocol)
AttributeError: 'module' object has no attribute 'SSLContext'

----------------------------------------------------------------------
Ran 5 tests in 0.002s

FAILED (errors=1)
```

Reply:
After Changes (including fix to TestUserUtils):



**Python 2.7.5**
```
# python --version
Python 2.7.5

# python  -m unittest -v  TestAmbariServer.TestAmbariServer
...
----------------------------------------------------------------------
Ran 125 tests in 5.541s

OK


# python  -m unittest -v  TestAmbariServer.TestAmbariServer^C
...
----------------------------------------------------------------------
Ran 5 tests in 0.001s

OK
```



**Python 2.7.15**
```
# python --version
Python 2.7.15

# python -m unittest -v  TestAmbariServer.TestAmbariServer
...
----------------------------------------------------------------------
Ran 125 tests in 5.556s

OK

# python  -m unittest -v  TestServerUtils.TestServerUtils
...
----------------------------------------------------------------------
Ran 5 tests in 0.001s

OK
```


**Python 2.7.6**
```
# ../Python-2.7.6/python --version
Python 2.7.6

# ../Python-2.7.6/python -m unittest -v  TestAmbariServer.TestAmbariServer
...
----------------------------------------------------------------------
Ran 125 tests in 5.562s

OK

# ../Python-2.7.6/python  -m unittest -v  TestServerUtils.TestServerUtils
...
----------------------------------------------------------------------
Ran 5 tests in 0.001s

OK
```

Reply:
Thanks!

BTW, I'd guess the reason for certain Python 2.7.5 having `SSLContext` is:

```
* Mon Jun 29 12:00:00 2015 Robert Kuska <rkuska@redhat.com> - 2.7.5-24
- Massive backport of ssl module from python3 aka PEP466
Resolves: rhbz#1111461
```